Add draft of Svkt extension (#1564)
riscv/riscv-isa-manual: src/supervisor.adoc

  {bits:  2, name: 'CBIE'},
  {bits:  1, name: 'CBCFE'},
  {bits:  1, name: 'CBZE'},

>   {bits:  1, name: 'SVKT'},

Reviewer: Since Sekt applies only to RV64 or larger, wouldn't it make sense to put `henvcfg. HUVKT` and `senvcfg.SVKT` at bit offset >= 32, like was done for PMM? This avoids consuming bits in the low word that are irrelevant for RV32.

Author: The ARC was not completely confident that we would never define an RV32 variant of this feature, hence locating the bit down here.

forthcoming Zicbom extension. Their allocations within `senvcfg` may
change prior to the ratification of that extension.

If the Svkt extension is implemented, the SVKT field affects the behavior of
instruction fetches and explicit memory accesses.
When SVKT=0, instruction fetches and explicit memory accesses proceed as
though the Svkt extension were not implemented.
When SVKT=1, an instruction fetch or explicit memory access with effective
privilege mode U or VU that attempts to access a virtual address {ge}
2^SXLEN-1^ raises a page-fault exception corresponding to the original access

> type.

Reviewer 2: Do we want more flexibility here? e.g. for 32-bit runtime. AFIAK, 32-bit Linux uses virtual addresses starting from `0xc0000000` as supervisor mapping. Different OSes may define it differently on 32-bit platform though, due to contested address space, e.g. windows fits into this half/half s/u division, IIRC.

Reviewer 3: Two comments here

1) I am not sure, if it's helpful for 32bit. 2 GB  (or 1GB when supervisor_addr_start=`0xc0000000`) is relatively small address space to have any significant benefit from KASLR.
So may be mandating only for 64 bit, probably makes it cleaner. 
Side note: Not that MSFT Windows should be of concern here but Windows 32bit also have an option to split address space in 3GB/1GB. Anyways 32bit windows is deprecated now.

2) `2^SXLEN-1` doesn't work when pointer masking is enabled because canonical checks are relaxed.

Instead of `2^SXLEN-1`, it should be

```
2^VABITS-1
VABITS = 39 if Sv39
VABITS = 48 if Sv48
VABITS = 57 if Sv57

If Bare mode then this bit has no effect?
```

Reviewer 3: Never mind about comment # 2.
Shouldn't be a problem, as long as effective address is calculated after masking is performed and generated address is canonicalized again.

Author: @AndybnACT We did discuss that, but the tentative plan is to support Svkt only for RV64 for the reason that @deepak0414 mentioned: KASLR is seemingly much less beneficial on RV32.  But if we did add an RV32 variant, we of course would design it so that it would be usable by the Linux kernel.  This would either mean using a 3G/1G split (most likely) or offering the choice between 3G/1G or 2G/2G (less likely).

@deepak0414 Right, re: # 2, the address transformation is applied before any virtual-address handling, and so there is no problem with pointer masking.

Author: I added a note about not having Svkt in RV32. 9a9ba46df4e7bc7f72288c33dc97e020324f59ff

coherent with store instructions that modify PTEs.
====

[[sec:svukte]]
== "Svukte" Extension for Address-Independent Latency of User-Mode Faults to Supervisor Addresses, Version 0.3

The Svukte extension provides a means to make user-mode accesses to supervisor

> memory raise page faults in constant time, mitigating attacks that attempt to

Reviewer 4: PR feedback from tech chairs: time invariant based on page table contents, not constant time of the whole exception handling.
```suggestion
memory raise page faults with timing independent of page table data, mitigating attacks that attempt to
```

Reviewer 5: I think the point is to not even look at the page table contents. The very act of accessing the page table will make it not constant time and could expose KASLR.

Author: I'm certainly open to other ways of phrasing this, but I agree with Ved that the proposed wording seems to unintentionally weaken the spec by allowing implementations to perform the first level of the page-table walk then discard the result.  (But I think the proposed wording does prohibit accessing subsequent levels of the page table.)

Unlike Ved, I haven't convinced myself that opens up a timing attack, but even so, it wasn't my intent.

Another option would be "with timing independent of the contents of `satp` or the page tables".

Another possibility is "in constant time" => "without accessing the page tables", but that has the downside of shifting the emphasis towards the implementation and away from the high-level intent.

Reviewer 4: Ooh I like that one: "with timing independent of the contents of satp or the page tables".

Author: Me too, but only if we can find a way to phrase it that includes all active `atp` registers.  "independent of the address-translation configuration and the page tables"?

Reviewer 5: Maybe we could leave the normative statement as is and add a short note to guide implementation with some text like "without accessing page tables or address translation caches" so that it provides enough hints that this fault is supposed to be a fault caused by agen, similar to an invalid VA or too-wide GPA case, and not a TLB/PWH caused fault?

Reviewer 5: Maybe we could leave the normative statement as is and add a short note to guide implementation with some text like "without accessing page tables or address translation caches" so that it provides enough hints that this fault is supposed to be a fault caused by agen, similar to an invalid VA or too-wide GPA case, and not a TLB/PWH caused fault?

Author: I can live with either my most recent proposal or Ved's suggestion.  Let's see if anyone else chimes in.

Author: I went with my version, but we can revisit this later if need be.

discover the supervisor software's address-space layout.

If the Svkt extension is implemented, the `senvcfg`.SVKT field is writable.
If the hypervisor extension is additionally implemented, the `henvcfg`.HUVKT
field is also writable.
See <<sec:senvcfg>> and <<sec:henvcfg>> for the definitions of those fields.

The Svkt extension depends on Sv39.

NOTE: Svkt is not defined for Sv32 because the small address space limits the
available entropy, reducing the effectiveness of address-space layout
randomization.
If an Sv32 variant were to be defined, it would need to account for the fact
that it is more common to reserve only the upper 1 GiB of the virtual-address

> space for for the operating system, leaving the lower 3 GiB for user

Author: thanks.

processes.

When `senvcfg`.UKTE=1, an instruction fetch or explicit memory access whose
effective privilege mode is U or VU is considered to be _Svukte-qualified_.
For any Svukte-qualified memory access, virtual addresses {ge} 2^SXLEN-1^ are

> considered to be invalid; hence, an Svukte-qualified access to such an address

Reviewer 6: > For any Svukte-qualified memory access, virtual addresses {ge} 2^SXLEN-1^ are
> considered to be invalid;

When CPU executes in VU-mode, do we also check 2^SXLEN-1 bit of virtual address (is invalid) in G-stage translation?

Author: I think the spec already says everything it needs to say. Svukte-qualified accesses are affected; the rest aren’t. Page-table walks aren’t Svukte-qualified.

Reviewer 6: Thanks for confirmation!

Reviewer 7: This doesn't explicitly say that the PTE.U bit is ignored, but it certainly implicitly does. I would make that explicit, and/or explain in section 10.31.1 that "The U bit indicates whether the page is accessible to user mode." [unless Svutke is enabled] or something

Author: The constraint on the implementation is much stricter than that. The implementation shouldn’t even perform the memory access corresponding to the PTE, and so it shouldn’t even be cognizant as to the value of PTE.U.

Don’t you think the next sentence in the spec (“The timing […] must be independent of the faulting VA”) gets this across? If not, how would you say it? Challenge: you cannot mention the PTE contents, given the machine’s behavior mustn’t depend on the PTE contents.

> If the hypervisor extension is additionally implemented, the `henvcfg`.HUKTE
> field is also writable.
> See <<sec:senvcfg>> and <<sec:henvcfg>> for the definitions of those fields.
> 
> The Svukte extension depends on Sv39.
> 
> NOTE: Svukte is not defined for Sv32 because the small address space limits the
> available entropy, reducing the effectiveness of address-space layout
> randomization.
> If an Sv32 variant were to be defined, it would need to account for the fact
> that it is more common to reserve only the upper 1 GiB of the virtual-address
> space for the operating system, leaving the lower 3 GiB for user
> processes.
> 
> When `senvcfg`.UKTE=1, an instruction fetch or explicit memory access whose
> effective privilege mode is U or VU is considered to be _Svukte-qualified_.
> For any Svukte-qualified memory access, virtual addresses {ge} 2^SXLEN-1^ are
> considered to be invalid; hence, an Svukte-qualified access to such an address
> raises a page-fault exception corresponding to the original access type.
> The timing of an instruction that raises an exception for this reason must be
> independent of the faulting virtual address.
> 
> NOTE: Since whether an instruction is Svukte-qualified depends on the _effective_
> privilege mode of the access, even some instructions executed in HS-mode or M-mode
> (e.g. HLV with `hstatus`.SPVP=0, or LW with `mstatus`.MPRV=1 and
> `mstatus`.MPP=U) are Svukte-qualified.
> 
> As described in <<sec:henvcfg>>, the `henvcfg`.HUKTE field, rather than the
> `senvcfg`.UKTE field, determines whether HLV, HLVX, and HSV instructions
> executed within U-mode are Svukte-qualified.
> 

Reviewer 4: Are there `HUKTE` fields in both `henvcfg` and `hstatus`?
```suggestion
If the hypervisor extension is additionally implemented, the `hstatus`.HUKTE
field is also writable.
See <<sec:senvcfg>> and <<sec:hstatus>> for the definitions of those fields.

The Svukte extension depends on Sv39.

NOTE: Svukte is not defined for Sv32 because the small address space limits the
available entropy, reducing the effectiveness of address-space layout
randomization.
If an Sv32 variant were to be defined, it would need to account for the fact
that it is more common to reserve only the upper 1 GiB of the virtual-address
space for the operating system, leaving the lower 3 GiB for user
processes.

When `senvcfg`.UKTE=1, an instruction fetch or explicit memory access whose
effective privilege mode is U or VU is considered to be _Svukte-qualified_.
For any Svukte-qualified memory access, virtual addresses {ge} 2^SXLEN-1^ are
considered to be invalid; hence, an Svukte-qualified access to such an address
raises a page-fault exception corresponding to the original access type.
The timing of an instruction that raises an exception for this reason must be
independent of the faulting virtual address.

NOTE: Since whether an instruction is Svukte-qualified depends on the _effective_
privilege mode of the access, even some instructions executed in HS-mode or M-mode
(e.g. HLV with `hstatus`.SPVP=0, or LW with `mstatus`.MPRV=1 and
`mstatus`.MPP=U) are Svukte-qualified.

As described in <<sec:hstatus>>, the `hstatus`.HUKTE field, rather than the
`senvcfg`.UKTE field, determines whether HLV, HLVX, and HSV instructions
executed within U-mode are Svukte-qualified.

```

Author: Editing error.  As the hypervisor chapter says, the bit lives in `hstatus`.

Author: Fixed.

coherent with store instructions that modify PTEs.
====

[[sec:svkt]]

> == "Svkt" Extension for Address-Independent Latency of User-Mode Translation of Supervisor Addresses, Version 0.2

Reviewer 4: The Faults are what is constant time, not the Translation (Table Walks).
```suggestion
== "Svkt" Extension for Address-Independent Latency of User-Mode Faults of Supervisor Addresses, Version 0.2
```

Author: The translation and the faulting are hard to disentangle, since it’s the variable latency of the translation that made the latency of the fault vary. But your version might make the intent clearer; we can try it out.

“faults to” rather than “faults of”, right?

Reviewer 4: Sure, "faults to".

Author: Done.

> The Svkt extension provides a means to make user-mode accesses to supervisor

Reviewer 4: Naming suggestion: it's not the whole translation that occurs with constant time, just the exceptions.
```suggestion
The Svkte extension provides a means to make user-mode accesses to supervisor
```

Author: Changing the name is a bit of a pain, so I'm going to wait to see if there are further suggestions, then pick the best.
